packages/x-data-grid/src/components/containers/GridOverlay.tsx

```diff
@@ -4,20 +4,71 @@ import clsx from 'clsx';
 import composeClasses from '@mui/utils/composeClasses';
 import { Theme, SxProps, styled } from '@mui/system';
 import type { DataGridProcessedProps } from '../../models/props/DataGridProps';
+import { minimalContentHeight } from '../../hooks/features/rows/gridRowsUtils';
 import { getDataGridUtilityClass } from '../../constants/gridClasses';
+import { useGridApiContext } from '../../hooks/utils/useGridApiContext';
+import { useGridSelector } from '../../hooks/utils/useGridSelector';
+import { gridDimensionsSelector } from '../../hooks/features/dimensions';
 import { useGridRootProps } from '../../hooks/utils/useGridRootProps';
+import { GridSlotsComponent } from '../../models';
 
 export type GridOverlayProps = React.HTMLAttributes<HTMLDivElement> & {
   sx?: SxProps<Theme>;
 };
 
+type GridLoadingOverlayVariant = 'circular-progress' | 'linear-progress' | 'skeleton';
+
+type GridOverlayType =
+  | keyof Pick<GridSlotsComponent, 'noRowsOverlay' | 'noResultsOverlay' | 'loadingOverlay'>
```

Comment:
duplication of types here, to get rid of Dependency cycle error.

Reply:
Moved to `useGridOverlays`